clients/src/main/java/org/apache/kafka/common/security/kerberos/KerberosName.java

```diff
@@ -0,0 +1,383 @@
+         * @throws BadFormatString
+         */
+        static String replaceParameters(String format,
+                                        String[] params) throws BadFormatString {
+            Matcher match = PARAMETERPATTERN.matcher(format);
+            int start = 0;
+            StringBuilder result = new StringBuilder();
+            while (start < format.length() && match.find(start)) {
+                result.append(match.group(1));
+                String paramNum = match.group(3);
+                if (paramNum != null) {
+                    try {
+                        int num = Integer.parseInt(paramNum);
+                        if (num < 0 || num > params.length) {
+                            throw new BadFormatString("index " + num + " from " + format +
+                                                      " is outside of the valid range 0 to " +
+                                                      (params.length - 1));
+                        }
+                        result.append(params[num]);
+                    } catch (NumberFormatException nfe) {
+                        throw new BadFormatString("bad format in username mapping in " +
+                                                  paramNum, nfe);
+                    }
+
+                }
+                start = match.end();
+            }
+            return result.toString();
+        }
+
+        /**
+         * Replace the matches of the from pattern in the base string with the value
+         * of the to string.
+         * @param base the string to transform
+         * @param from the pattern to look for in the base string
+         * @param to the string to replace matches of the pattern with
+         * @param repeat whether the substitution should be repeated
+         * @return
+         */
+        static String replaceSubstitution(String base, Pattern from, String to,
+                                          boolean repeat) {
+            Matcher match = from.matcher(base);
+            if (repeat) {
+                return match.replaceAll(to);
+            } else {
+                return match.replaceFirst(to);
+            }
+        }
+
+        /**
+         * Try to apply this rule to the given name represented as a parameter
+         * array.
+         * @param params first element is the realm, second and later elements are
+         *        are the components of the name "a/b@FOO" -> {"FOO", "a", "b"}
+         * @return the short name if this rule applies or null
+         * @throws IOException throws if something is wrong with the rules
+         */
+        String apply(String[] params) throws IOException {
+            String result = null;
+            if (isDefault) {
+                if (defaultRealm.equals(params[0])) {
+                    result = params[1];
+                }
+            } else if (params.length - 1 == numOfComponents) {
```

Comment:
It's a bit weird that # of params doesn't match numOfComponents. Could you add a comment?